### ISSUE TITLE
JIT SystemSegmentProvider leaks memory if an inconveniently-sized allocation is repeatedly requested then released

### ISSUE THREAD

**Issue by jdmpapin** [eclipse-openj9/openj9#13353]
When allocating more than the default segment size, but less than the system segment size, `J9::SystemSegmentProvider` will [allocate within a system segment](https://github.com/eclipse-openj9/openj9/blob/ca4973a2b0d9414a726e826c1e5d5506761c395d/runtime/compiler/env/SystemSegmentProvider.cpp#L238) by [bumping a pointer](https://github.com/eclipse-openj9/openj9/blob/ca4973a2b0d9414a726e826c1e5d5506761c395d/runtime/compiler/env/J9SegmentProvider.cpp#L43-L45). When the allocation is later released, it is [split up into default-sized segments](https://github.com/eclipse-openj9/openj9/blob/ca4973a2b0d9414a726e826c1e5d5506761c395d/runtime/compiler/env/SystemSegmentProvider.cpp#L200-L212). I expect that the intention was to add these to the free list (`_freeSegments`), but I don't see where that happens. But even if they are/were added to the free list, these smaller segments are no longer usable for allocations of the original size

[The default segment size is 64kB](https://github.com/eclipse-openj9/openj9/blob/ca4973a2b0d9414a726e826c1e5d5506761c395d/runtime/compiler/control/CompilationThread.cpp#L7959) and [the system segment size is 16MB](https://github.com/eclipse-openj9/openj9/blob/ca4973a2b0d9414a726e826c1e5d5506761c395d/runtime/compiler/control/CompilationThread.cpp#L7961) (also [for the `J9SegmentCache`](https://github.com/eclipse-openj9/openj9/blob/ca4973a2b0d9414a726e826c1e5d5506761c395d/runtime/compiler/control/CompilationThread.cpp#L3935)). So any allocation between 64kB and 16MB will trigger this behaviour. For example, the following loop leaks memory even though the `StackMemoryRegion` is destroyed every iteration and "releases" the allocated memory
```
for (int i = 0; i < 8192; i++)
   {
   TR::StackMemoryRegion region(*comp()->trMemory());
   region.allocate((1 << 16) + 1);
   }
```

**Comment by dsouzai** [comment]
Yeah I'm not sure why, when we carve up the segment in `J9::SystemSegmentProvider::release`, those new segments aren't added to `_freeSegments`; probably could just be fixed with a two-line PR, something like:

```
diff --git a/runtime/compiler/env/SystemSegmentProvider.cpp b/runtime/compiler/env/SystemSegmentProvider.cpp
index 28905b2fd9..6b33df31ab 100644
--- a/runtime/compiler/env/SystemSegmentProvider.cpp
+++ b/runtime/compiler/env/SystemSegmentProvider.cpp
@@ -203,7 +203,8 @@ J9::SystemSegmentProvider::release(TR::MemorySegment & segment) throw()
          {
          try
             {
-            createSegmentFromArea(defaultSegmentSize(), static_cast<uint8_t *>(oldSegmentArea) + (defaultSegmentSize() * i));
+            TR::MemorySegment &temp = createSegmentFromArea(defaultSegmentSize(), static_cast<uint8_t *>(oldSegmentArea) + (defaultSegmentSize() * i));
+            _freeSegments.push_back(TR::ref(temp));
             }
          catch (...)
             {
```

In terms of why the carving happens at all, part of it might be because it may have seemed unlikely there would be many allocations greater than 64KiB (though that is just my guess) and partly because the complexity of having multiple sized free lists wasn't something that was desired<sup>1</sup>.

<hr/>

1. A free list would be needed here because the segment is smaller than the system segment and so it can't be released back to VM/OS.

**Comment by jdmpapin** [comment]
That would be good, but it wouldn't solve the problem in the example loop. If we change 8192 to `n`, a proper fix would allow that loop to run in constant space. Probably the way to go IMO is to have only one size threshold, i.e. equate the "default" and "system" segment sizes. Then there would only be regular reusable segments (which get put on the free-list) and large segments (which are released immediately). I say this because I don't think we should get into the business of efficiently reusing freed space of arbitrary size - that's getting into general-purpose allocator territory, and we might as well just call the existing one

**Comment by gaolibj** [comment]
Is this problem solved?  We found a high CPU issue in allocateNewSegment. Not sure if it's related.

**Comment by jdmpapin** [comment]
No, this isn't fixed yet, unfortunately
